### ISSUE TITLE
[BUG] Failed to determine a type for ZodEffects - transform

### ISSUE THREAD

**Issue by bdashore3** [rhinobase/hono-openapi#3]
Hi, found an issue when calling the `openAPISpecs` function if one route uses a schema that has a `.transform()`.

Here's the error that shows up when I startup my application:
```
error: Uncaught (in promise) Error: Failed to determine a type for ZodEffects - transform at createSchema. Please change the 'effectType' to 'input', wrap it in a ZodPipeline or assign it a manual 'type'.
    at createManualOutputTransformSchema (file:///C:/Users/kingbri/AppData/Local/deno/npm/registry.npmjs.org/zod-openapi/3.3.0/dist/components.chunk.mjs:186:11)
    at createTransformSchema (file:///C:/Users/kingbri/AppData/Local/deno/npm/registry.npmjs.org/zod-openapi/3.3.0/dist/components.chunk.mjs:160:15)
    at createSchemaSwitch (file:///C:/Users/kingbri/AppData/Local/deno/npm/registry.npmjs.org/zod-openapi/3.3.0/dist/components.chunk.mjs:1198:12)
    at createNewSchema (file:///C:/Users/kingbri/AppData/Local/deno/npm/registry.npmjs.org/zod-openapi/3.3.0/dist/components.chunk.mjs:1248:18)
    at createSchemaOrRef (file:///C:/Users/kingbri/AppData/Local/deno/npm/registry.npmjs.org/zod-openapi/3.3.0/dist/components.chunk.mjs:1344:10)
    at createSchemaObject (file:///C:/Users/kingbri/AppData/Local/deno/npm/registry.npmjs.org/zod-openapi/3.3.0/dist/components.chunk.mjs:1348:18)
    at createSchema (file:///C:/Users/kingbri/AppData/Local/deno/npm/registry.npmjs.org/zod-openapi/3.3.0/dist/components.chunk.mjs:1353:18)
    at createSchema (file:///C:/Users/kingbri/AppData/Local/deno/npm/registry.npmjs.org/zod-openapi/3.3.0/dist/index.mjs:41:18)
    at Object.builder (file:///C:/Users/kingbri/AppData/Local/deno/npm/registry.npmjs.org/hono-openapi/0.1.4/zod.esm.js:1:164)
    at resolver (file:///C:/Users/kingbri/AppData/Local/deno/npm/registry.npmjs.org/hono-openapi/0.1.4/zod.esm.js:1:331)
```

To make this error not occur, remove the `.transform()` call or comment out the `/openapi` route. I've included a test server below. Change Deno to whatever runtime you use.

```ts
import { Hono } from "hono";
import { describeRoute, openAPISpecs } from "hono-openapi";
import { resolver, validator as zValidator } from "hono-openapi/zod";
import { z } from "zod";
import { apiReference } from "@scalar/hono-api-reference";

const app = new Hono();

const testSchema = z.object({
    input: z.string(),
    aliased_input: z.string().optional(),
    dynamic_input: z.string().optional(),
})
.transform((obj) => {
    return {
        ...obj,
        aliased_input: obj.aliased_input ?? obj.dynamic_input ?? "Hello!"
    }
})

const extendedSchema = z.object({
    extended_input: z.string()
}).and(testSchema);


const responseSchema = z.object({
    finish: z.string()
})

app.post(
    "/",
    describeRoute({
        description: "Says hello",
        responses: {
            200: {
                description: "Success",
                content: {
                    "application/json": {
                        schema: resolver(responseSchema)
                    }
                }
            }
        }
    }),
    zValidator("json", extendedSchema),
    (c) => {
        console.log(c.req.valid('json'));
        return c.json(responseSchema.parse({ finish: "Hello!" }))
    }
)

app.get(
    "/openapi",
    openAPISpecs(app)
);

app.get(
    "/reference",
    apiReference({
        spec: {
            url: "/openapi",
        },
    }),
);

Deno.serve(app.fetch)
```

**Comment by samchungy** [comment]
Hey there, this is because you're using a ZodEffect which makes it difficult for us to to introspect the schema you are trying to create https://github.com/samchungy/zod-openapi#zod-effects

**Comment by bdashore3** [comment]
Thanks for the quick response!

Yes, setting `effectType` to `input` fixed that problem, but another one popped up. When chaining multiple `.and`s, the documentation nests them together instead of splitting them apart. See the image below for more information.

See the following example:
```ts
import { Hono } from "hono";
import { describeRoute, openAPISpecs } from "hono-openapi";
import { resolver, validator as zValidator } from "hono-openapi/zod";
import { z } from "zod";
import { apiReference } from "@scalar/hono-api-reference";

import "zod-openapi/extend";

const app = new Hono();

const firstSchema = z.object({
    first_input: z.string(),
    aliased_input: z.string().optional(),
    dynamic_input: z.string().optional(),
})
.transform((obj) => {
    return {
        ...obj,
        aliased_input: obj.aliased_input ?? obj.dynamic_input ?? "Hello!"
    }
})
.openapi({
    description: "First Schema",
    effectType: "input"
})

const secondSchema = z.object({
    second_input: z.string(),
    aliased_input: z.string().optional(),
    dynamic_input: z.string().optional(),
})
.transform((obj) => {
    return {
        ...obj,
        aliased_input: obj.aliased_input ?? obj.dynamic_input ?? "Hello!"
    }
})
.openapi({
    description: "Second Schema",
    effectType: "input"
})

const extendedSchema = z.object({
    extended_input: z.string()
})
.and(firstSchema)
.and(secondSchema)
.openapi({
    description: "Extended Schema"
});

const responseSchema = z.object({
    finish: z.string()
})

app.post(
    "/",
    describeRoute({
        description: "Says hello",
        responses: {
            200: {
                description: "Success",
                content: {
                    "application/json": {
                        schema: resolver(responseSchema)
                    }
                }
            }
        }
    }),
    zValidator("json", extendedSchema),
    (c) => {
        return c.json(responseSchema.parse({ finish: "Hello!" }))
    }
)

app.get(
    "/openapi",
    openAPISpecs(app)
);

app.get(
    "/reference",
    apiReference({
        spec: {
            url: "/openapi",
        },
    }),
);

Deno.serve(app.fetch)
```

![image](https://github.com/user-attachments/assets/d36ef63d-be88-440d-91f7-7f972385a2b2)



With Hono's official zod openapi middleware (and how the request is correctly represented in JSON):

```ts
import { z } from "@hono/zod-openapi";
import { OpenAPIHono, createRoute } from "@hono/zod-openapi";
import { apiReference } from "@scalar/hono-api-reference";

const app = new OpenAPIHono();

const firstSchema = z.object({
    first_input: z.string(),
    aliased_input: z.string().optional(),
    dynamic_input: z.string().optional(),
})
.transform((obj) => {
    return {
        ...obj,
        aliased_input: obj.aliased_input ?? obj.dynamic_input ?? "Hello!"
    }
})
.openapi({
    description: "First Schema",
})

const secondSchema = z.object({
    second_input: z.string(),
    aliased_input: z.string().optional(),
    dynamic_input: z.string().optional(),
})
.transform((obj) => {
    return {
        ...obj,
        aliased_input: obj.aliased_input ?? obj.dynamic_input ?? "Hello!"
    }
})
.openapi({
    description: "Second Schema",
})

const extendedSchema = z.object({
    extended_input: z.string()
})
.and(firstSchema)
.and(secondSchema)
.openapi({
    description: "Extended Schema"
});

const responseSchema = z.object({
    finish: z.string()
});

const testRoute = createRoute({
    method: "post",
    path: "/",
    request: {
        body: {
            content: {
                "application/json": {
                    schema: extendedSchema
                }
            }
        }
    },
    responses: {
        200: {
            content: {
                'application/json': {
                    schema: responseSchema,
                },
            },
            description: 'Retrieve the user',
        },
    }
})

app.openapi(
    testRoute,
    (c) => {
        return c.json(responseSchema.parse({ finish: "Hello!" }))
    }
)

app.doc("/openapi", {
    openapi: "3.0.0",
    info: {
        version: "0.0.1",
        title: "YALS",
    },
});

app.get(
    "/reference",
    apiReference({
        spec: {
            url: "/openapi",
        },
    }),
);

Deno.serve(app.fetch)
```

![image](https://github.com/user-attachments/assets/16ee9a0c-4623-4bdb-a39b-9de8fc4c21e8)

**Comment by samchungy** [comment]
Actually this might be a repo specific issue, the library should be setting `schemaType: 'input'` for request schemas.

**Comment by samchungy** [comment]
I can't understand that image at all, could you provide the JSON outputs?

**Comment by bdashore3** [comment]
Sure.

Here's the JSON output from this library (the first code example above)

```json
{
  "openapi": "3.0.0",
  "info": {
    "title": "Hono Documentation",
    "description": "Development documentation",
    "version": "0.0.0"
  },
  "paths": {
    "/": {
      "post": {
        "responses": {
          "200": {
            "description": "Success",
            "content": {
              "application/json": {
                "schema": {
                  "type": "object",
                  "properties": {
                    "finish": {
                      "type": "string"
                    }
                  },
                  "required": [
                    "finish"
                  ]
                }
              }
            }
          }
        },
        "operationId": "postIndex",
        "description": "Says hello",
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "allOf": [
                  {
                    "allOf": [
                      {
                        "type": "object",
                        "properties": {
                          "extended_input": {
                            "type": "string"
                          }
                        },
                        "required": [
                          "extended_input"
                        ]
                      },
                      {
                        "type": "object",
                        "properties": {
                          "first_input": {
                            "type": "string"
                          },
                          "aliased_input": {
                            "type": "string"
                          },
                          "dynamic_input": {
                            "type": "string"
                          }
                        },
                        "required": [
                          "first_input"
                        ],
                        "description": "First Schema"
                      }
                    ]
                  },
                  {
                    "type": "object",
                    "properties": {
                      "second_input": {
                        "type": "string"
                      },
                      "aliased_input": {
                        "type": "string"
                      },
                      "dynamic_input": {
                        "type": "string"
                      }
                    },
                    "required": [
                      "second_input"
                    ],
                    "description": "Second Schema"
                  }
                ],
                "description": "Extended Schema"
              }
            }
          }
        }
      }
    }
  },
  "components": {
    "schemas": {}
  }
}
```

And here's a JSON output using [Hono's official openapi](https://www.npmjs.com/package/@hono/zod-openapi) middleware

```json
{
  "openapi": "3.0.0",
  "info": {
    "title": "Hono Documentation",
    "description": "Development documentation",
    "version": "0.0.0"
  },
  "components": {
    "schemas": {},
    "parameters": {}
  },
  "paths": {
    "/": {
      "post": {
        "requestBody": {
          "content": {
            "application/json": {
              "schema": {
                "allOf": [
                  {
                    "type": "object",
                    "properties": {
                      "extended_input": {
                        "type": "string"
                      }
                    },
                    "required": [
                      "extended_input"
                    ]
                  },
                  {
                    "type": "object",
                    "properties": {
                      "first_input": {
                        "type": "string"
                      },
                      "aliased_input": {
                        "type": "string"
                      },
                      "dynamic_input": {
                        "type": "string"
                      }
                    },
                    "required": [
                      "first_input"
                    ],
                    "description": "First Schema"
                  },
                  {
                    "type": "object",
                    "properties": {
                      "second_input": {
                        "type": "string"
                      },
                      "aliased_input": {
                        "type": "string"
                      },
                      "dynamic_input": {
                        "type": "string"
                      }
                    },
                    "required": [
                      "second_input"
                    ],
                    "description": "Second Schema"
                  }
                ],
                "description": "Extended Schema"
              }
            }
          }
        },
        "responses": {
          "200": {
            "description": "Retrieve the user",
            "content": {
              "application/json": {
                "schema": {
                  "type": "object",
                  "properties": {
                    "finish": {
                      "type": "string"
                    }
                  },
                  "required": [
                    "finish"
                  ]
                }
              }
            }
          }
        }
      }
    }
  }
}
```

Apologies for the length. I can attach these as files if necessary.

**Comment by samchungy** [comment]
Ah I see, it's still technically correct, but rather they're flattening it further. I can add that.

**Comment by bdashore3** [comment]
Will there need to be a new version of zod-openapi for this change to take effect?

**Comment by samchungy** [comment]
> Will there need to be a new version of zod-openapi for this change to take effect?

It'll be in the next version yep, when I decide to release.

**Comment by MathurAditya724** [comment]
Hey @samchungy! It's awesome to see you contributing here – thank you for helping out ❤️. 

Regarding `schemaType`, I'm setting `"input"` for all request validation handlers. You can check out how this is applied at this line: https://github.com/rhinobase/hono-openapi/blob/main/packages/core/src/zod.ts#L105. This metadata is consistently passed to all request validation handlers, including query, params, json, etc.

I ran into a similar issue during testing, and I resolved it by adding a `.pipe` method after `.transform`. You can find more details on how this works here: https://zod.dev/?id=pipe.

**Comment by samchungy** [comment]
Hmmm does that get passed to the `createSchema` options?

<img width="413" alt="image" src="https://github.com/user-attachments/assets/8ddca47c-994a-4980-a8da-8d8ad52ed69d">

**Comment by samchungy** [comment]
> Will there need to be a new version of zod-openapi for this change to take effect?

This is in 4.0.0, alongside a breaking change. Sorry not sorry 😅

**Comment by MathurAditya724** [comment]
> Hmmm does that get passed to the `createSchema` options?

Yes, Over here - https://github.com/rhinobase/hono-openapi/blob/main/packages/core/src/openapi.ts#L58

**Comment by MathurAditya724** [comment]
This is how my schema validation ended up like - 
```ts
export const booleanStringValidation = z
  .string()
  .trim()
  .toLowerCase()
  .refine((val) => VALID_BOOLEAN_VALUES.includes(val), "Invalid boolean value")
  .transform((val) => val === "true" || val === "1")
  .pipe(z.boolean());
```

**Comment by samchungy** [comment]
> > Hmmm does that get passed to the `createSchema` options?
> 
> Yes, Over here - [`main`/packages/core/src/openapi.ts#L58](https://github.com/rhinobase/hono-openapi/blob/main/packages/core/src/openapi.ts#L58)

I think you're explicitly mapping the options here though right?

https://github.com/rhinobase/hono-openapi/blob/37147e7a403395c871fff746eb254bc434d48a8b/packages/core/src/zod.ts#L16-L20

**Comment by MathurAditya724** [comment]
ahh! 👀 yeah that is my bad, I will correct that thanks!

**Comment by MathurAditya724** [comment]
Just released a new version `hono-openapi@0.1.5`, which seems to resolve the issue at my end. @samchungy Thanks for pointing it out.
